### PR TITLE
ci: use download-artifact @ 4.1.1

### PR DESCRIPTION
**Pull request by imjasonh**

### CODE REVIEW

**Comment by imjasonh** [comment]
Disabled the ABI check requirement since it seems to be skipped when there are no packages built. I'll flip it back to required after this lands.

File: .github/workflows/build-world.yaml [hunk 1]
```diff
@@ -55,7 +55,7 @@ jobs:
         # TODO: See how big these get, maybe we only upload failures and shorten the retention, or throw them in GCS
       - name: Upload build logs
         if: always()
-        uses: actions/upload-artifact@v3
+        uses: actions/upload-artifact@5d5d22a31266ced268874388b861e4b58bb5c2f3 # v4.3.1
```
[the remark this answers is not in the document]
Comment (imjasonh) [reply .github/workflows/build-world.yaml:58]: In general it's just a best practice that we should be more diligent and consistent about.

If it's referenced by tag someone at GH could push over the tag and we'd immediately start getting new (possibly malicious) behavior. If it's by commit SHA they can't. The `# 4.3.1` comment helps dependabot send automated PRs to keep it up to date with new commit SHAs.

[the remark this answers is not in the document]
Comment (smoser) [reply .github/workflows/build-world.yaml:58]: that was what I suspected. thank for clarifying.

**Review by smoser — APPROVED** [review]
thank you.

**Review by developer-guy — APPROVED** [review]

**Review by dlorenc — APPROVED** [review]

**Pull request merged**